### DISCUSSION TITLE
SuperTextField bugfixes

### DISCUSSION THREAD

**Discussion by matthew-carroll** [Flutter-Bounty-Hunters/super_editor#527] [Milestones]
At the moment, `SuperTextField` works on both desktop and mobile, but it has a number of bugs.

This milestone covers a number of bugfixes for `SuperTextField`.

 * #505
 * #504
 * #506
 * #459
 * #529
 * #508 
 * #444 
 * #460 
 * #383 
 * #549 
 * #620 
 * #621 
 * #622 
 * #278 
 * #275 
 * #523 
 * #619 
 * #583 

## Funders
Turtle, Superlist

**Comment by matthew-carroll** [comment]
@venkatd - would you like to list the priority order of the bugfixes from your perspective?

**Reply by venkatd** [reply]
@matthew-carroll sure. I can take a closer look tomorrow but can start with the top two that stand out

## Issues in order of priority 
https://github.com/superlistapp/super_editor/issues/505
https://github.com/superlistapp/super_editor/issues/508
https://github.com/superlistapp/super_editor/issues/504
https://github.com/superlistapp/super_editor/issues/506
https://github.com/superlistapp/super_editor/issues/459
https://github.com/superlistapp/super_editor/issues/529

**Reply by venkatd** [reply]
@matthew-carroll Should I maintain the above as a prioritized backlog of issues Turtle is willing to sponsor?

**Reply by matthew-carroll** [reply]
@venkatd I think we should choose a specific set of tickets for this milestone and then work towards completion. Otherwise the milestone may never end. We can always file another milestone after this one.

Does this list still represent your priorities for text field bug fixes?

**Reply by matthew-carroll** [reply]
It looks like #508 #509 #510 are also filed against super text field. Should those be included?

**Reply by venkatd** [reply]
@matthew-carroll yes that's right I'd keep the issues above as the right order of priorities. I've added #508 to the list but have left the other two out because they are low priority compared to the list above.

**Reply by matthew-carroll** [reply]
What do you think about adding some testing tools to this milestone? I think that some of the bugs may be the result of a lack of testing infrastructure. I'm thinking about.

Easy text field content asserts:
```
// The "{", "}" characters represent bold. The "|" represents the caret.
// The meaning of surrounding characters, like "{ }" still needs some thought.
// We'll want them to be configurable, but also we want the standard case to
// be easy to parse. Maybe we parse Markdown as the standard format...
expect(find.byType(SuperTextField), hasText("Hello, {world!}|"));
```

Extensions for `WidgetTester`:
```
await tester.tapAtTextOffset(5);

await tester.doubleTapAtTextOffset(5);

await tester.dragExpandSelectionTo(10);

await tester.pressRight();

await tester.pressShiftRight();

await tester.pressShiftAltRight();
```

A fake `TextLayout` implementation so that some multi-line behaviors can be tested without figuring out exactly how Flutter's Ahem font lays out text in a widget test.

What do you think?

**Reply by venkatd** [reply]
@matthew-carroll that sounds good to me.

I don't understand the full details of it, but if it helps stabilize things, let's go for it!
